test/backend/AnnotationUserStateTestSuite.scala

```diff
@@ -0,0 +1,104 @@
+      assert(renderedUserState.treeIds == Seq(1, 2))
+      assert(renderedUserState.treeVisibilities == Seq(false, true))
+      assert(renderedUserState.activeNodeId == Some(5))
+      assert(renderedUserState.treeGroupIds == Seq(1))
+      assert(renderedUserState.treeGroupExpandedStates == Seq(true))
+    }
+
+    "be rendered into new skeleton user state correctly for userB (owner)" in {
+      val renderedUserState =
+        renderSkeletonUserStateIntoUserState(dummySkeletonWithUserState, "userB", "userB")
+      assert(renderedUserState.treeIds == Seq(1, 2))
+      assert(renderedUserState.treeVisibilities == Seq(true, true))
+      assert(renderedUserState.activeNodeId == Some(2))
+      assert(renderedUserState.treeGroupIds == Seq.empty)
+      assert(renderedUserState.treeGroupExpandedStates == Seq.empty)
+    }
+
+    "be rendered into new skeleton user state correctly for userC (no user state present for them)" in {
+      val renderedUserState =
+        renderSkeletonUserStateIntoUserState(dummySkeletonWithUserState, "userC", "userB")
+      assert(renderedUserState.treeIds == Seq(1, 2))
+      assert(renderedUserState.treeVisibilities == Seq(true, true))
+      assert(renderedUserState.activeNodeId == Some(2))
+      assert(renderedUserState.treeGroupIds == Seq.empty)
+      assert(renderedUserState.treeGroupExpandedStates == Seq.empty)
+    }
+
+  }
+
+  "volume user states merging" should {
+    "respect id mapping" in {
+      val tracingAUserStates = Seq(
+        VolumeTracingDefaults
+          .emptyUserState("userA")
+          .copy(
+            segmentIds = Seq(1),
+            segmentVisibilities = Seq(true),
+            segmentGroupIds = Seq(1),
+            segmentGroupExpandedStates = Seq(true)
+          ))
+      val tracingBUserStates = Seq(
+        VolumeTracingDefaults
+          .emptyUserState("userA")
+          .copy(
+            segmentIds = Seq(1),
+            segmentVisibilities = Seq(false),
+            segmentGroupIds = Seq(1),
+            segmentGroupExpandedStates = Seq(false)
+          ))
+
+      val segmentIdMapB = Map((1L, 2L))
+      val mergedUserStates = mergeVolumeUserStates(tracingAUserStates,
+                                                   tracingBUserStates,
+                                                   groupMappingA = (groupId: Int) => groupId + 5,
+                                                   segmentIdMapB,
+                                                   Map.empty,
+                                                   Map.empty)
+      assert(
+        mergedUserStates == Seq(
+          VolumeTracingDefaults
+            .emptyUserState("userA")
+            .copy(
+              segmentIds = Seq(1L, 2L),
+              segmentVisibilities = Seq(true, false),
+              segmentGroupIds = Seq(6, 1),
+              segmentGroupExpandedStates = Seq(true, false),
+            )))
```

Comment:
I think I am misunderstanding something here. Doesn't `mergeVolumeUserStates` merge two volume user states of the same underlying volume data? In that case this test reads like segment id 1 is suddenly mapped to segment id 2 which would have a very new meaning. Is the same ensured to be applied to the bucket data of segmentation of the newly merged volume data? What if the segment with id 1 in tracing B was never modified and only loaded into the segment list? What happens in this case? To my understanding the previously segment 1 loaded in the tracingB does now have segmentId 2 and is falsely referring to a different segment id in the fallback data?

Reply:
Absolutely right, you can only merge annotations that were created on the same dataset.